### ISSUE TITLE
[no-extra-parens] incorrect behaviour when using a parenthesised array type in a function call generic

### ISSUE THREAD

**Issue by movpushmov** [typescript-eslint/typescript-eslint#4077]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**
1. create React project with typescript
2. insert rule into your .eslintrc.json file
3. insert code into your functional component
4. see error

```JSON
{
  "rules": {
    "@typescript-eslint/no-extra-parens": ["error"]
  }
}
```

```TS
const [state, setSomeState] = useState<(string | number)[]>(
	someVar ??
        anotherSomeVar ??
        anotherSomeSomeVar
)
```

**Expected Result**
This code must be marked typescript-eslint as correct

**Actual Result**
Eslint throws me this error: ESLint: Unnecessary parentheses around expression.(@typescript-eslint/no-extra-parens)

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `5.2.0` |
| `@typescript-eslint/parser`        | `5.2.0` |
| `TypeScript`                       | `4.4.3` |
| `ESLint`                           | `7.32.0` |
| `node`                             | `14.18.1` |

**Comment by movpushmov** [comment]
Update:

Well, it's looks like issue with typing state. This is workaround:

type Key = string | number

```TS
// this throws an error
const [state, setSomeState] = useState<(string | number)[]>(
	someVar ??
        anotherSomeVar ??
        anotherSomeSomeVar
)

// this doesn't throw error
const [state, setSomeState] = useState<Key[]>(
	someVar ??
        anotherSomeVar ??
        anotherSomeSomeVar
)
```

**Comment by armano2** [comment]
simplified test cases

```
fn<(number)[]>(true)
fn<(number | boolean)[]>(true)
```

**Comment by movpushmov** [comment]
@armano2  @bradzacher any changes?

**Comment by bradzacher** [comment]
With any issue in opened in this project - it either has a visible progress in the form of an attached PR, or it has no progress.

We are a community run project. The **volunteer** maintainers spend most of their time triaging issues and reviewing PRs. This means that most issues *will not progress* unless a member of the community steps up and **champions** it.

If this issue is important to you - consider being that champion.

If not - please just subscribe to the issue and wait patiently.
Commenting asking for status updates does not bump issue priority in any way and just serves to spam ***everyone*** subscribed to the issue.

**Comment by 6XGate** [comment]
This also affects functions passed to generic functions, for example: `ref<(item: unknown) => string>(() => '')`

**Comment by auvred** [comment]
Fixed in https://github.com/typescript-eslint/typescript-eslint/pull/7375
